### PR TITLE
Optimizer Rule for Disjoint Smart Graphs.

### PR DESCRIPTION
**Pull request by markuspf**
Make an optimizer rule that supports disjoint smart edge collections.

### CODE REVIEW

**Comment by markuspf** [comment]
Note that most of this PR is happening in enterprise code https://github.com/arangodb/enterprise/pull/438

**Comment by markuspf** [comment]
http://jenkins.arangodb.biz:8080/job/arangodb-matrix-pr/9544/

**Comment by markuspf** [comment]
http://jenkins.arangodb.biz:8080/job/arangodb-matrix-pr/9549/

**Comment by markuspf** [comment]
http://jenkins.arangodb.biz:8080/job/arangodb-matrix-pr/9556/

**Comment by markuspf** [comment]
http://jenkins.arangodb.biz:8080/job/arangodb-matrix-pr/9560/

**Comment by markuspf** [comment]
http://jenkins.arangodb.biz:8080/job/arangodb-matrix-pr/9583/

File: arangod/Aql/OptimizerRules.cpp [hunk 1]
```diff
@@ -3903,13 +3903,261 @@ void arangodb::aql::scatterInClusterRule(Optimizer* opt, std::unique_ptr<Executi
...
+      createKeys = true;
+    } break;
+    case ExecutionNode::TRAVERSAL: {
+      auto traversalNode = ExecutionNode::castTo<TraversalNode const*>(node);
+      TRI_ASSERT(traversalNode->isDisjoint());
+      collection = traversalNode->collection();
+      inputVariable = traversalNode->inVariable();
+      // TODO:
+      // If the traversal node uses a constant start vertex, then this will be
+      // nullptr, hence we'll have to stunt around this.
```
Comment (mchacki) [inline arangod/Aql/OptimizerRules.cpp:3972]:
```suggestion
```
done in another PR I have just reviewed ;)

File: arangod/Aql/OptimizerRules.cpp [hunk 2]
```diff
@@ -3903,13 +3903,261 @@ void arangodb::aql::scatterInClusterRule(Optimizer* opt, std::unique_ptr<Executi
...
+      createKeys = false;
+    } break;
+    case ExecutionNode::SHORTEST_PATH: {
+      auto shortestPathNode = ExecutionNode::castTo<ShortestPathNode const*>(node);
+      TRI_ASSERT(shortestPathNode->isDisjoint());
+      collection = shortestPathNode->collection();
+      inputVariable = shortestPathNode->startInVariable();
+      // TODO:
+      // If the traversal node uses a constant start vertex, then this will be
+      // nullptr, hence we'll have to stunt around this.
```
Comment (mchacki) [inline arangod/Aql/OptimizerRules.cpp:3995]:
```suggestion
```
Done in another PR.

File: arangod/Aql/OptimizerRules.cpp [hunk 3]
```diff
@@ -3903,13 +3903,261 @@ void arangodb::aql::scatterInClusterRule(Optimizer* opt, std::unique_ptr<Executi
...
+  auto const parents = at->getParents();
+  auto const deps = at->getDependencies();
+
+  // This transforms `parents[0] -> node -> deps[0]` into `parents[0] -> deps[0]`
+  plan.unlinkNode(at, true);
+
+  // create, and register a distribute node
+  DistributeNode* distNode = createDistributeNodeFor(plan, at);
+  TRI_ASSERT(distNode != nullptr);
+  distNode->addDependency(deps[0]);
```
Comment (mchacki) [inline arangod/Aql/OptimizerRules.cpp:4061]:
```suggestion
TRI_ASSERT(deps.size() == 1);
  distNode->addDependency(deps[0]);
```

I think this code is only valid if you have EXACTLY one dependency.

I am rather certain that we right now only create those plans at this place, however this assertion helps in the future.

File: arangod/Aql/OptimizerRules.cpp [hunk 4]
```diff
@@ -3903,13 +3903,261 @@ void arangodb::aql::scatterInClusterRule(Optimizer* opt, std::unique_ptr<Executi
...
+  // insert another remote node
+  remoteNode =
+      plan.createNode<RemoteNode>(&plan, plan.nextId(), vocbase, "", "", "");
+  remoteNode->addDependency(at);
+
+  // insert a gather node matching the distribute node
+  auto* gatherNode = createGatherNodeFor(plan, distNode);
+  gatherNode->addDependency(remoteNode);
+
+  // Song and dance to deal with at being the root of a plan or a subquery
```
Comment (mchacki) [inline arangod/Aql/OptimizerRules.cpp:4088]:
```suggestion
TRI_ASSERT(parents.size() < 2);
  // Song and dance to deal with at being the root of a plan or a subquery
```

This code only works if do not have multiple parents of this node.
I think right now, we do not create such kind of plans here, but this is future proof
